README.md
# CiteIt Web Citation Service

The CiteIt Web Citation Service allows web authors
to **demonstrate the context of their quotations** by helping them display the
500 characters of context before and after the quote they are citing.

## View Demo: ##
Demo site with video:
  * https://www.CiteIt.net/
  
Sample API Calls:
  * https://www.citeit.net/sample-code/  

## Tools Used ##
  * **API Server**: [Python](https://www.python.org/) [Flask](http://flask.pocoo.org/) Server
  * **Database**: [SQL Alchemy](https://www.sqlalchemy.org/), 
                  ([Postgres](https://www.postgresql.org/) or Other Database)
  * **[jQuery Library](https://github.com/CiteIt/citeit-jquery)**: custom function to wrap &lt;blockquote&lt; and &lt;q&gt; tags
  * **[Wordpress Plugin](https://github.com/CiteIt/citeit-wordpress)**: 
    * adds buttons to Wordpress editor for blockquote and pre tags.
    * the working version was written before  Gutenberg.
    * (looking for help creating a Gutenberg version of the plugin)
    * Submits pages to API.
    
  * **Dependencies**: 
    * [Google Diff Match Patch](https://code.google.com/archive/p/google-diff-match-patch/):
        fuzzy text matching algorithm
    * [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/): used to convert html 
        to text and locate the canonical urls within html
    * [Flask](https://palletsprojects.com/p/flask/)
    * [SQL Alchemy](https://www.sqlalchemy.org/)
    * [boto3](https://github.com/boto/boto3)
    * Python [Requests](http://docs.python-requests.org/en/master/): HTTP for humans
    * [ftfy](http://ftfy.readthedocs.io/en/latest/): convert bad unicode to good unicode
              


## How CiteIt Works:

1. An author notifies the CiteIt Web Service that they have created a new
citation by sending a HTTP POST request with the URL of the author's page 
to api.citeit.net.

1. The CiteIt Web Service retrieves the author's page and locates all the
citations within the document, saving the urls and the text of each citation.

1. The CiteIt web service retrieves the pages for each cited URL and
compares the cited source text with the text from the author's citation.

1. If the service finds a match, it calculates the 500 characters before
and after each quotation.

1. The CiteIt web service saves the 500 characters before and after each
quotation to a json file and uploads each json file to the Amazon S3 Cloud.

1. After the author installs either the CiteIt Wordpress plugin
or the CiteIt jQuery library on their site, the plugin instructs subsequent
visitors to load the contextual json files from the Amazon S3 Cloud.

1. The CiteIt jQuery library, loads the contextual data from the json file
into hidden &lt;div&gt; elements created in the author's page, which display when
a reader clicks on an arrow above or below the quotation.

## Setup:

### Flask:
cd app/

cp settings-default.py settings.py

(populate settings.py with own passwords)

export FLASK_APP=app.py

set FLASK_RUN_PORT=80

python3 -m flask run --host=0.0.0.0 --port=80

### Docker:
docker build -t citeit_webservice:latest .

docker tag ce0bdceb4791 citeit/citeit_webservice:v0.4.2
docker push citeit/citeit_webservice:v0.4.2

docker run -p 80:80 -e AMAZON_ACCESS_KEY=password -e AMAZON_SECRET_KEY=password citeit/citeit_webservice:latest

## Inspiration:
I got this idea in 2015, while I was writing an article about hypertext pioneer
[Ted Nelson](https://en.wikipedia.org/wiki/Ted_Nelson).

  * https://www.openpolitics.com/articles/ted-nelson-philosophy-of-hypertext.html

Like the Web, CiteIt still follows the "[Worse is Better](https://www.dreamsongs.com/RiseOfWorseIsBetter.html)" approach because I don't have the technical ability or resources to implement the 
Ted's "Right Thing".

I hope CiteIt allows more people to get a glimpse of Ted's original proposal and 
inspires other programmers to built towards that vision.

-**[Tim Langeman](https://www.openpolitics.com/tim/)**

Akron, PA (USA)

## License ##
I want this form of expanding contextual citation to **become as widely adopted as possible**, so I'm releasing it under 
the **[MIT License](https://opensource.org/licenses/MIT)**.  I encourage you to build on top of my code.  You don't have to 
contribute back to this project, but there are advantages to cooperation.


## History:
Earlier versions of this concept were written for:
1. PHP: https://github.com/timlangeman/quote-context
1. Django: https://github.com/neotext/neotext-django-server


changelog.txt
Aug 17, 2020: 0.4.7
  * add permalink to blockquote and q tags to display context on non canonical pages (the home page, etc

July 30, 2020: 0.4.6
  * add "Right and Left Quotation" to Unicode Escape characters
  * change style: q-tag link: double underline, remove border from quote arrows

July 25, 2020: 0.4.5
  * update dependency: bump lodaash from 4.17.15 to 44.17.19 in /lib/link-to-text-fragment
  * format q-tag links and popup window
  * add error handliniking to YouTube embed: url_parsed.params.start

July 6, 2020: 0.4.4
  * add Logic to Expand: YouTube Vimeo, and Soundcloud
  * fix Quote-Context Arrows: remove border

Jun 27, 2020: 0.4.3
  * change version to 3-part number
  * convert UTF-16 to UTF-8

June 22, 2020: 0.4.2. xx
  * change webservice submission url: http://api.citeit.net/v0.4/url/
  * aadd forge-sha256 build
  * update jQuery version to 3.4.1
  * update forge-sha256

June 19, 2020: 0.4.1 (0.64)
  * update jsVideoURLParser    
  * preserve html in TinyMCE editor quote selection
	

June 18, 2020 (0.63)
 * format and document code
 

June 13, 2020 (0.62)
  * add function isWordpressPreview() to CiteIt-quote-context:
    - detect if current URL is in preview mode
    - remove querystring if in "preview" mode
  * rename existing function names: use camel case
  
  https://github.com/CiteIt/citeit-wordpress/issues/11 
  

Jan 29, 2020 (0.61)
  * cleanup CiteIt-quote-context  
  * add <a class="quote_arrows">
  
  * switch toggle_quote to reference arrow anchor
    rather than parent div

  * CSS
    - add <a class=‘quote_arrows’ to support styling
    - change display to “block” to get rotate to work
    - remove link outline from arrows
    - add transform support for legacy browsers  
  * make contextual popup appear on top: z-index: 9999
  * rename neotext -> citeit


Jan 25, 2020 (0.60)
  * quote-context.php: register style, remove text/css attrib from style sheets
  * enque style shee  
  * fix validatio errors
  * add toggleClass("rotated180");

Jan 18, 2020:
  * Escape Text using Unicode Codepoints 
  * switch endpoint to: http://api.citeit.net
  
  
Dec 20, 2020:
  * Fix CSS: Inline Popup hides behind other Page Elements (#19)
  * change z-index of jQuery Popup using .ui-front class
  * this fixes #19
  
  
See more: https://github.com/CiteIt/citeit-wordpress/commits/master

app/settings.py
import os
import sys
import logging


logging.basicConfig(filename="citeit-webservice.log", 
                    format='%(asctime)s %(message)s', 
                    filemode='w') 

logger=logging.getLogger() 
logger.setLevel(logging.DEBUG) 


# Webservice is versioned in JSON URL
VERSION_NUM = os.getenv('VERSION_NUM', '0.4')

# example JSON URL:
# https://read.citeit.net/quote/sha256/0.4/d5/d588c1c9c4acfcd254acc4033b7888e98f21e426214b21f0e07673664e328e39.json

logger=logging.getLogger()
logger.setLevel(logging.DEBUG) 

PDF_ENABLED = False

JSON_FILE_PATH = '/tmp/'   # in Lambda, you need to save to /tmp folder
SQLALCHEMY_DATABASE_URI = 'postgresql://username:password@localhost/database_name'

SAVE_DOWNLOADS_TO_FILE = True
#############################################################################################################################
# Certain Unicode characters can cause problems with the hash
# Remove the following Unicode code points from Hash

URL_ESCAPE_CODE_POINTS = set ([
  10, 20, 160
])

# Remove the following Unicode code points from Hash
TEXT_ESCAPE_CODE_POINTS = set ( [
  2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 39, 96, 133, 160, 173, 699, 700, 701, 702, 703, 712, 713, 714, 715, 716, 717, 718, 719, 732, 733, 750, 757, 5760, 6158, 8203, 8192, 8193, 8194, 8195, 8196, 8197, 8198, 8199, 8200, 8201, 8202, 8204, 8205, 8211, 8212, 8213, 8216, 8217, 8219, 8220, 8221, 8226,   8232, 8233,  8239, 8287, 8288, 12288, 65279
])


ESCAPE_SPECIAL_CHARS = [
  # Example of characters that could be escaped: (currently commented out)

  #'\u00e2\u0080\u009d',  # "â"
  #'\u00e2\u0080\u009c',  # LEFT DOUBLE QUOTATION MARK
  #'\u00e2\u0080\u0093',  # DASH
  #'\u00e2\u0080\u00a6',
  #'\u00e2\u0080\u0098',  # LEFT SINGLE QUOTATION
  #'\u00e2\u0080\u0099',  # RIGHT SINGLE QUOTATION
]
#######################################################################################
# Set number of Quote lookups to make simulaneously from a given URL:
# Usage in: app/lib/citeit_quote_context/url.py
# pool = Pool(processes=settings.NUM_DOWNLOAD_PROCESSES)

NUM_DOWNLOAD_PROCESSES = 5

# aws_setting is stored in grandparent path
sys.path.append(os.path.abspath('../../'))

try:
    # Are Settings supplied by an included Amazon file?
    import aws_settings  # path: (../../aws_settings.py)

    AMAZON_ACCESS_KEY = aws_settings.AMAZON_ACCESS_KEY
    AMAZON_SECRET_KEY = aws_settings.AMAZON_SECRET_KEY
    AMAZON_S3_BUCKET = aws_settings.AMAZON_S3_BUCKET
    AMAZON_S3_ENDPOINT = aws_settings.AMAZON_S3_ENDPOINT
    AMAZON_REGION_NAME = aws_settings.AMAZON_REGION_NAME

except ImportError:
    # Use Docker Settings
    AMAZON_ACCESS_KEY = os.getenv('AMAZON_ACCESS_KEY', '')    # 'ABCDEFGH123456789'
    AMAZON_SECRET_KEY = os.getenv('AMAZON_SECRET_KEY', '')    # 'alksdfj;2452lkjr;ajtsaljgfslakjfgassgf'
    AMAZON_S3_BUCKET = os.getenv('AMAZON_S3_BUCKET', '')      # 'read.citeit.net'
    AMAZON_S3_ENDPOINT = os.getenv('AMAZON_S3_ENDPOINT', '')  # 's3.amazonaws.com'
    AMAZON_REGION_NAME = os.getenv('AMAZON_REGION_NAME', '')  # 'us-east-1'

###############################################################################################################################
# Default to AWS File Input, if it exists. Get Amazon Settings from Docker Input,
#
# Docker params are passed in at the command-line :
#   docker run -p 80:80
#		-e AMAZON_ACCESS_KEY= **************
#		-e AMAZON_SECRET_KEY= **************
#		citeit/citeit_webservice:latest
#
